cleverhans/attacks_eager.py

```diff
@@ -0,0 +1,260 @@
+from abc import ABCMeta
+import numpy as np
+from six.moves import xrange
+import warnings
+import collections
+import tensorflow as tf
+from distutils.version import LooseVersion
+import cleverhans.utils as utils
+from cleverhans.attacks import AttackBase
+from cleverhans.model_eager import Model_Eager
+
+_logger = utils.create_logger("cleverhans.attacks_eager")
+
+if LooseVersion(tf.__version__) < LooseVersion('1.8.0'):
+    error = ('For eager execution',
+             'use Tensorflow version greather than 1.8.0.')
+    raise ValueError(error)
+
+
+class AttackTFE(AttackBase):
+    """
+    Abstract base class for all attack classes.
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, model, dtypestr='float32'):
+        """
+        :param model: An instance of the cleverhans.model.Model_Eager class.
+        :param back: The backend to use. Inherited from AttackBase class.
+        :param dtypestr: datatype of the input data samples and crafted
+                        adversarial attacks.
+        """
+        # Validate the input arguments.
+        if dtypestr != 'float32' or dtypestr != 'float64':
+            raise ValueError("Unexpected input for argument dtypestr.")
+        import tensorflow as tf
+        tfe = tf.contrib.eager
+        self.tf_dtype = tf.as_dtype(dtypestr)
+        self.np_dtype = np.dtype(dtypestr)
+
+        if not isinstance(model, Model_Eager):
+            raise ValueError("The model argument should be an instance of"
+                             " the cleverhans.model.Model_Eager class.")
+        # Prepare attributes
+        self.layer_names = []
+
+    def get_or_guess_labels(self, x, kwargs):
```

Comment:
What is specific to eager in this method? It looks very similar to the one in tf_attacks.py

Reply:
I see your point, do you suggest overloading Attack class definitions ?? (i'm open to this ... )
my thoughts -- 
in attacks_tf.py all the functions are defined as part of graph execution and run through session.
in eager mode they run on the fly.
code logic is similar by design, (they are meant to do same thing in different context.)

Reply:
Yes, that's exactly my thoughts, the code is the same except for gradient the tape and execution, so inheriting from tf_attack so that the eager code is minimal is a much more impressive demonstration of its simplicity than having a lot of duplicate which makes more intimidating.

Reply:
Sure, makes sense, will do that ...